google/cloud/pubsub_v1/publisher/flow_controller.py

```diff
@@ -0,0 +1,145 @@
+# limitations under the License.
+
+import logging
+import threading
+import warnings
+
+from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.publisher import exceptions
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FlowController(object):
+    """A class used to control the flow of messages passing through it.
+
+    Args:
+        settings (~google.cloud.pubsub_v1.types.PublishFlowControl):
+            Desired flow control configuration.
+    """
+
+    def __init__(self, settings):
+        self._settings = settings
+
+        self._message_count = 0
+        self._total_bytes = 0
+
+        # The lock is used to protect the internal state (message and byte count).
+        self._operational_lock = threading.Lock()
+
+        # The condition for blocking the flow if capacity is exceeded.
+        self._has_capacity = threading.Condition(lock=self._operational_lock)
+
+    def add(self, message):
+        """Add a message to flow control.
+
+        Adding a message updates the internal load statistics, and an action is
+        taken if these limits are exceeded (depending on the flow control settings).
+
+        Args:
+            message (:class:`~google.cloud.pubsub_v1.types.PubsubMessage`):
+                The message entering the flow control.
+
+        Raises:
+            :exception:`~pubsub_v1.publisher.exceptions.FlowControlLimitError`:
+                If adding a message exceeds flow control limits and the desired
+                action is :attr:`~google.cloud.pubsub_v1.types.LimitExceededBehavior.ERROR`.
+        """
+        if self._settings.limit_exceeded_behavior == types.LimitExceededBehavior.IGNORE:
+            return
+
+        with self._operational_lock:
+            self._message_count += 1
+            self._total_bytes += message.ByteSize()
+
+            if not self._is_overflow():
+                return
+
+            # We have an overflow, react.
+            if (
+                self._settings.limit_exceeded_behavior
+                == types.LimitExceededBehavior.ERROR
+            ):
+                msg = (
+                    "Flow control limits exceeded "
+                    "(messages: {} / {}, bytes: {} / {})."
+                ).format(
+                    self._message_count,
+                    self._settings.message_limit,
+                    self._total_bytes,
+                    self._settings.byte_limit,
+                )
+                error = exceptions.FlowControlLimitError(msg)
+
+                # Raising an error means rejecting a message, thus we need to deduct
+                # the latter's contribution to the total load.
+                self._message_count -= 1
+                self._total_bytes -= message.ByteSize()
+                raise error
+
+            assert (
+                self._settings.limit_exceeded_behavior
+                == types.LimitExceededBehavior.BLOCK
+            )
+
+            while self._is_overflow():
```

Comment:
I'm not sure this is going to have the desired effect. Imagine I set flow control to allow ten messages through and I call publish 30 times. self._message_count will increment to 30. Now, when a message finishes, we won't call self._has_capacity.notify_all() because self._is_overflow() is still true. In fact, we will never call notify_all because the number of outstanding messages will never drop below 20.

Reply:
This issue only occurs when the behavior is set to types.LimitExceededBehavior.BLOCK. If it's set to LimitExceededBehavior.ERROR, message_count/total_bytes are not increased if the publish is rejected. (See line 89-90.)

A simple way to fix this is to reduce the message_count/total_bytes before blocking, and to add them again after the wait returns, but before the is_overflow check is done again. This way, we don't wait while using phantom quota.

Perhaps a more elegant way is to change _is_overflow to take the new message's size as a parameter. Then, we can check if the it would overflow before modifying the variables. Then, we'd only need to increment count/size at the end of the function (end of the `with` block.)

Reply:
> Now, when a message finishes, we won't call self._has_capacity.notify_all() because self._is_overflow() is still true. In fact, we will never call notify_all because the number of outstanding messages will never drop below 20.

By design, `notify_all()` acts on a "negative front", i.e. when the state changes from "overflow" to "not overflow anymore". Every release decreases the load, and while a single released message will indeed not call `notify_all()`, the latter will be called after releasing enough messages.

> I'm not sure this is going to have the desired effect. Imagine I set flow control to allow ten messages through and I call publish 30 times. self._message_count will increment to 30.

That's a bug, thanks for catching it! The message count should never increase to 30, but should instead block when 11th message arrives. That message should only be accepted, i.e. added to the load, after releasing at least one message.

Essentially the case covered in the ERROR path (the load adjustment @pradn mentioned), but forgot to add the same to the BLOCK path.

> Perhaps a more elegant way is to change _is_overflow to take the new message's size as a parameter. Then, we can check if the it would overflow before modifying the variables.

Indeed. :+1: Fixing the BLOCK path would result in too many load adjustments scattered around the method, hindering readability, thus the overflow logic will probably be refactored.

Reply:
@kamalaboulhosn @pradn Thank you both for the valuable comments!

I actually thought we indeed have a bug in the BLOCK path, but thinking through it again makes me believe we got it right after all. At least I was not able to reproduce the supposed buggy behavior with a few additional tests I ran locally, but will nevertheless check the existing FlowController tests once again to see if they somehow missed something.

Reply:
I figured out a simple fix would possibly lead to thread starving - a thread adding a large message would "never" detect enough capacity, since other threads posting smaller message would keep stealing the chunks of available capacity from it.

Solved by by keeping track of waiting threads and greedily distributing any freed-up capacity among them in FIFO order.